filter/ip2location/filterip2location.go

```diff
@@ -0,0 +1,235 @@
+				if firstTime {
+					firstTime = false
+				} else {
+					fc.reloadFile()
+				}
+			case <-fc.watcher.Events:
+				timer.Reset(pauseDelay)
+			case err := <-fc.watcher.Errors:
+				goglog.Logger.Errorf("ip2location: %s", err.Error())
+			case <-fc.ctx.Done():
+				return
+			}
+		}
+	}()
+}
+
+// InitHandler initialize the filter plugin
+func InitHandler(ctx context.Context, raw *config.ConfigRaw) (config.TypeFilterConfig, error) {
+	conf := DefaultFilterConfig()
+	err := config.ReflectConfig(raw, &conf)
+	if err != nil {
+		return nil, err
+	}
+
+	conf.ctx = ctx
+
+	// open database
+	conf.db, err = ip2location.OpenDB(conf.DBPath)
+	if err != nil {
+		return nil, err
+	}
+
+	// init cache
+	conf.cache, err = lru.New(conf.CacheSize)
+	if err != nil {
+		return nil, err
+	}
+
+	goglog.Logger.Info("ip2location fsnotify initialized for", conf.DBPath)
+
+	// init fsnotify
+	conf.watcher, err = fsnotify.NewWatcher()
+	if err != nil {
+		goglog.Logger.Errorf("ip2location failed to init watcher, %s", err.Error())
+	}
+	err = conf.watcher.Add(conf.DBPath)
+	if err != nil {
+		goglog.Logger.Errorf("ip2location failed to add file: %s", err.Error())
+	}
+	conf.initFsnotifyEventHandler()
+
+	var cidrs []string
+	if len(conf.PrivateNet) > 0 {
+		cidrs = conf.PrivateNet
+	} else {
+		cidrs = DefaultCIDR
+	}
+
+	for _, cidr := range cidrs {
+		_, privateCIDR, err := net.ParseCIDR(cidr)
+		if err != nil {
+			return nil, err
+		}
+		conf.privateCIDRs = append(conf.privateCIDRs, privateCIDR)
+	}
+
+	return &conf, nil
+}
+
+// not_supported is copied from ip2location and is the field entry for each field that is not supported by the current database
+const not_supported string = "This parameter is unavailable for selected data file. Please upgrade the data file."
+
+// Event the main filter event
+func (f *FilterConfig) Event(ctx context.Context, event logevent.LogEvent) (logevent.LogEvent, bool) {
+	ipstr := event.GetString(f.IPField)
+	if ipstr == "" {
+		// Passthru if empty
+		return event, false
+	}
+	ip := net.ParseIP(ipstr)
+	if ip == nil {
+		return event, false
+	}
+	if f.SkipPrivate && f.privateIP(ip) {
+		// Passthru
+		return event, false
+	}
+	var record *ip2location.IP2Locationrecord
+	if c, ok := f.cache.Get(ipstr); ok {
+		record = c.(*ip2location.IP2Locationrecord)
+	} else {
+		f.dbMtx.RLock()
+		r2, err := f.db.Get_all(ipstr)
+		f.dbMtx.RUnlock()
+		record = &r2
+		if err != nil {
+			if f.QuietFail {
+				goglog.Logger.Error(err)
+			}
+			event.AddTag(ErrorTag)
+			return event, false
+		}
+		f.cache.Add(ipstr, record)
+	}
+	if record == nil {
+		event.AddTag(ErrorTag)
+		return event, false
+	}
+
+	m := map[string]interface{}{
+		"country_code": record.Country_short,
+		"country_name": record.Country_long,
+	}
+	if record.City != not_supported {
+		m["city_name"] = record.City
+		m["region_name"] = record.Region
+	}
+	if record.Isp != not_supported {
+		m["ISP"] = record.Isp
+	}
+	if record.Latitude != 0 || record.Longitude != 0 {
```

Comment:
Would it be more robust if changing `||` to `&&`?

Reply:
If any of the numbers are non-zero then there is a coordinate where the other number is zero. It can theoretically happen.

I tried to browse through the DB5LITE CSV file to find any occurrences of this but could not find any. They claim the paid database has better accuracy (meaning more unique coordinates.) The closest entry I found to 0 was "England","Bluntisham","52.354790","0.008540".

I suggest we leave it as it is. From the CVS I only find (0,0) used on invalid IP adresses.